ultravox/inference/infer.py

```diff
@@ -68,55 +88,53 @@ def infer(
         output_tokens = output.sequences[0][input_len:]
         output_text = self.tokenizer.decode(output_tokens, skip_special_tokens=True)
         output_len = len(output_tokens)
-
         if self.conversation_mode:
-            past_messages = copy.deepcopy(extended_sample.messages)
-            audio_token_len = (
-                0 if "audio_token_len" not in inputs else inputs["audio_token_len"][0]
+            audio_token_len = inputs.get("audio_token_len", [0])[0]
+            past_messages = self._build_past_messages(
+                extended_sample.messages, audio_token_len, output_text
             )
-            if audio_token_len > 0:
-                user_content = past_messages[-1]["content"]
-                if user_content.count("<|audio|>") != 1:
-                    raise ValueError(
-                        f"Expected 1 audio placeholder, found {user_content.count('<|audio|>')}"
-                    )
-                past_messages[-1]["content"] = user_content.replace(
-                    "<|audio|>", self.tokenizer.eos_token * audio_token_len
-                )
-            past_messages.append({"role": "assistant", "content": output_text})
             self.update_conversation(past_messages, output.past_key_values)
-
         return base.VoiceOutput(output_text, input_len, output_len)
 
-    # streaming is not supported in conversation mode yet, to be implemented
     def infer_stream(
         self,
         sample: datasets.VoiceSample,
         max_tokens: Optional[int] = None,
         temperature: Optional[float] = None,
     ) -> base.InferenceGenerator:
-        inputs = self._dataproc(sample)
+        extended_sample = self._get_sample_with_past(sample)
+        inputs = self._dataproc(extended_sample)
         input_tokens = inputs["input_ids"].shape[1]
         decode_kwargs = {"skip_special_tokens": True}
         streamer = transformers.TextIteratorStreamer(
             self.tokenizer, skip_prompt=True, decode_kwargs=decode_kwargs
         )
 
-        thread_args = (
-            inputs,
-            max_tokens,
-            temperature,
-            streamer,
-        )
-        thread = threading.Thread(target=self._generate, args=thread_args)
+        def thunk(q: queue.Queue):
+            result = self._generate(
+                inputs, max_tokens, temperature, streamer, self.past_key_values
+            )
+            q.put(result)
+
+        result_queue: queue.Queue = queue.Queue()
+        thread = threading.Thread(target=thunk, args=(result_queue,))
         thread.start()
-        output_tokens = 0
+        output_text = ""
+        output_token_len = 0
         for chunk in streamer:
             if chunk:
+                output_text += chunk
+                output_token_len += 1
```

Comment:
Not really relevant to PR, but is every token representable in text?

I have a vague memory of some binary/unicode tokens (e.g. Chinese?) could not be decoded into actual text with `tiktoken` until their "pair" or "suffix" was completed.

I could be wrong of course and even then it's a very unlikely event.

Reply:
Is the concern that we shouldn't be appending to `output_text`, or that `output_token_len` will be off?

Reply:
Not sure what would happen in this case tbh. Since a chunk was already generated, then I guess the only possibility is that the `output_token_len` will be off, but I'm not really sure if this can even happen.
Just thought to raise this in case we come across it later on. Feel free to resolve.